strawberryfields/utils/post_processing.py

```diff
@@ -0,0 +1,341 @@
+            the variance for
+
+    Returns:
+        float: the variance from the samples
+    """
+    return _samples_variance(homodyne_samples, modes)
+
+
+def _samples_expectation(samples, modes):
+    """Utility function for calculating the expectation value of samples from
+    multiple modes.
+
+    Args:
+        homodyne_samples (ndarray): the homodyne samples with a shape of
+            (shots, modes)
+        modes (Sequence): a flat sequence containing indices of modes to get
+            the expectation value for
+
+    Returns:
+        float: the expectation value from the samples
+    """
+    _check_samples(samples)
+
+    num_modes = samples.shape[1]
+
+    if modes is None:
+        modes = np.arange(num_modes)
+    else:
+        _check_modes(samples, modes)
+
+    product_across_modes = _product_for_modes(samples, modes)
+    expval = product_across_modes.mean()
+    return expval
+
+
+def _samples_variance(samples, modes):
+    """Utility function for calculating the variance of samples from multiple
+    modes.
+
+    Args:
+        homodyne_samples (ndarray): the homodyne samples with a shape of
+            (shots, modes)
+        modes (Sequence): a flat sequence containing indices of modes to get
+            the variance for
+
+    Returns:
+        float: the variance from the samples
+    """
+    _check_samples(samples)
+
+    num_modes = samples.shape[1]
+
+    if modes is None:
+        modes = np.arange(num_modes)
+    else:
+        _check_modes(samples, modes)
+
+    product_across_modes = _product_for_modes(samples, modes)
+    variance = product_across_modes.var()
+    return variance
+
+
+def all_fock_probs_pnr(photon_number_samples):
+    r"""The Fock state probabilities for the specified modes.
+
+    Measured modes that are not specified are being traced over. If either all
+    the modes or no modes were specified, the marginal probabilities are
+    returned.
+
+    **Example:**
+
+    .. code-block:: python
+
+        samples = np.array([[2, 0],
+                            [2, 2],
+                            [2, 0],
+                            [0, 0]])
+
+    Getting the probabilities of all the Fock states based on these PNR samples:
+
+    >>> probs = all_fock_probs_pnr(samples)
+
+    2 out of 4 shots resulted in the ``(2, 0)`` state. We can check if the
+    probability for this state is correct by indexing into ``probs``:
+
+    >>> probs[(2, 0)]
+    0.5
+
+    We can further check the entire array of probabilities:
+    >>> probs
+    [[0.25 0.   0.  ]
+    [0.   0.   0.  ]
+    [0.5  0.   0.25]]
+
+    Args:
+        photon_number_samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+
+    Returns:
+        array: array of dimension :math:`\underbrace{D\times D\times D\cdots\times D}_{\text{num modes}}`
+            containing the Fock state probabilities, where :math:`D` is the Fock basis cutoff truncation
```

Comment:
How is the cutoff computed? Is the cutoff simply `max_val`? If so, it should probably explain that in the docstring.

Reply:
I think so.

Reply:
Valid point, thanks! Added `The Fock basis cutoff truncation is determined by the maximum Fock state with a non-zero probability.` to the docstring.